docs: Provide blog post on PredictKube scaler (#650)
kedacore/keda-docs: content/blog/2022-02-09-predictkube-scaler.md

So, Dysnix started thinking about predictive autoscaling. To scale in time, you have to scale beforehand—or else, you'll have to deal with another problem that's familiar to each cloud inhabitant. The overprovision of resources is used to cover any traffic changes and results in bills and thrills connected with unexpected expenses. Can this process be more controllable?

And when it comes to deciding how to wrap up this solution, Dysnix decided to rely on KEDA as it is the most universal and applicable component for application autoscaling in Kubernetes.

KEDA is being used as a component on the client side of PredictKube that is responsible for transferring requests and scaling replicas.

## Dysnix's PredictKube integrates with KEDA

Dysnix has built [PredictKube](https://predictkube.com/), a solution that can be used as a KEDA scaler that is responsible for resource balancing, and an AI model that has learned to react proactively to patterns of traffic activity, to help with both in-time scaling and solving the problem of overprovision.

> ![](/img/blog/predictkube-scaler/overprovision.png)

Reviewer: This image is way too big, can we resize it please?

See https://deploy-preview-650--keda.netlify.app/blog/2022-02-09-predictkube-scaler/

Author: the reason is that the output of the scaledobject command is very wide. Can we exclude it from the article or should I fix the css by adding scrolling to the code and a fixed width?

Author: @tomkerkhove I have removed commands outputs and checked locally - everything looks well

With this tool, Dysnix wants to decrease costs on the projects, analyze the data about traffic more efficiently, use cloud resources more responsibly, and build infrastructures that are &quot;greener&quot; and more performative (with fewer downtimes and delays) than others.

## How does PredictKube work?

PredictKube works in two parts:

1. **On the KEDA side**
   The interface connects via API to the data sources about your traffic. PredictKube uses [Prometheus](https://prometheus.io/)—the industry standard for storing metrics. There, it anonymizes the data about the traffic on the client's side before sending it to the API, where the model then works with information that is completely impersonal.
2. **On the AI model side**
   Next, it is linked with a prediction mechanism—the AI model starts to get data about things that happen in your cloud project. Unlike standard rules-based algorithms such as Horizontal Pod Autoscaling (HPA), PredictKube uses Machine Learning models for time series data predicting, like CPU or RPS metrics.

The more data you can provide to it from the start, the more precise the prediction will be. The 2+ weeks data will be enough for the beginning.

> ![](/img/blog/predictkube-scaler/diagram.svg)

Reviewer: Same here, let's resize and use a PNG/JPG

Author: converted to pnd

To check the configuration and status of the scaling created in the previous step, use the following command:

```bash
$ kubectl get scaledobject example
NAME      SCALETARGETKIND                    SCALETARGETNAME   MIN   MAX   TRIGGERS      AUTHENTICATION                         READY   ACTIVE   FALLBACK   AGE
example   apps.kruise.io/v1alpha1.CloneSet   example           5     30    predictkube   keda-trigger-auth-predictkube-secret   True    True     False      7d22h
```

To get stats to use for the scaling, use the following command:

```bash
$ kubectl get hpa example                                                                                                                                                                                                         (pancake-eu/bsc)
NAME               REFERENCE          TARGETS            MINPODS   MAXPODS   REPLICAS   AGE
keda-hpa-example   CloneSet/example   943400m/1k (avg)   5         30        5          7d2h
```
Now you can look at how scaling works at a graph in your visualization tool. This is an example of a graph Dysnix gets in one of their projects after using PredictKube:

> ![](/img/blog/predictkube-scaler/grafana.png)

Reviewer: Same here

+++
title = "Introducing PredictKube - an AI-based predictive autoscaler for KEDA made by Dysnix"

> date = 2022-02-09

Reviewer: If we want to publish on Monday, let's update the date
```suggestion
date = 2022-02-14
```

Author: updated, thank you!

author = "Daniel Yavorovych (Dysnix), Yuriy Khoma (Dysnix), Zbynek Roubalik (KEDA), Tom Kerkhove (KEDA)"
aliases = [

> "/blog/introducing-predictkube-an-ai-based-predictive-autoscaler-for-keda-made-by-dysnix"

Reviewer: Let's simplify the URL a bit
```suggestion
"/blog/introducing-predictkube-for-keda-by-dysnix"
```

+++
title = "Introducing PredictKube - an AI-based predictive autoscaler for KEDA made by Dysnix"
date = 2022-02-09
author = "Daniel Yavorovych (Dysnix), Yuriy Khoma (Dysnix), Zbynek Roubalik (KEDA), Tom Kerkhove (KEDA)"
aliases = [
"/blog/introducing-predictkube-an-ai-based-predictive-autoscaler-for-keda-made-by-dysnix"
]
+++

## Introducing PredictKube—an AI-based predictive autoscaler for KEDA made by Dysnix

> [Dysnix](https://dysnix.com/) has been working with big high-loaded projects for a long time,

Reviewer 2: ```suggestion
[Dysnix](https://dysnix.com/) has been working with high-traffic backend systems for a long time,
```

and the efficient scaling demand is what their team comes across each day.

> The engineers have understood that dealing with traffic fluctuations and preparations of infrastructure by bare hand is

Reviewer 2: ```suggestion
The engineers have understood that manually dealing with traffic fluctuations and preparations of infrastructure is
```

> totally inefficient as you need to deploy more resources _before_ the traffic increases,

Reviewer 2: ```suggestion
inefficient because you need to deploy more resources _before_ the traffic increases,
```

> not at the moment the event happens. It is too late to scale when the big traffic wave has already come.

Reviewer 2: ```suggestion
not at the moment the event happens. This strategy is problematic for two reasons: first, because it's often too late to scale when traffic has already arrived and second, resources will be overprovisioned and idle during the times that traffic isn't present.
```

So, Dysnix started thinking about predictive autoscaling. To scale in time, you have to scale beforehand—or else,
you'll have to deal with another problem that's familiar to each cloud inhabitant.
The overprovision of resources is used to cover any traffic changes and results in bills and thrills connected with unexpected expenses.
Can this process be more controllable?

And when it comes to deciding how to wrap up this solution, Dysnix decided to rely on KEDA as it is the most
universal and applicable component for application autoscaling in Kubernetes.

KEDA is being used as a component on the client side of PredictKube that is responsible for transferring requests

> and scaling replicas.

Reviewer 2: I think that these 3 paragraphs aren't needed if you mention overprovisioning above (in my previous comment).

Author: Thanks for your suggestion, Aaron, but we'd like to leave a part about the KEDA role for PredictKube.
